modules/graceful/manager_windows.go

```diff
@@ -0,0 +1,147 @@
+
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+// This code is heavily inspired by the archived gofacebook/gracenet/net.go handler
+
+package graceful
+
+import (
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/debug"
+)
+
+var WindowsServiceName = "gitea"
+
+const (
+	hammerCode       = 128
+	hammerCmd        = svc.Cmd(hammerCode)
+	acceptHammerCode = svc.Accepted(hammerCode)
+)
+
+
+type gracefulManager struct {
+	isChild                bool
+	lock                   *sync.RWMutex
+	state                  state
+	shutdown               chan struct{}
+	hammer                 chan struct{}
+	terminate              chan struct{}
+	runningServerWaitGroup sync.WaitGroup
+	createServerWaitGroup  sync.WaitGroup
+	terminateWaitGroup     sync.WaitGroup
+}
+
+func newGracefulManager() *gracefulManager {
+	manager := &gracefulManager{
+		isChild: false,
+		lock:    &sync.RWMutex{},
+	}
+	manager.createServerWaitGroup.Add(numberOfServersToCreate)
+	manager.Run()
+	return manager
+}
+
+func (g *gracefulManager) Run() {
+	g.setState(stateRunning)
+	if skip, _ := strconv.ParseBool(os.Getenv("SKIP_MINWINSVC")); skip {
+		return
+	}
+	run := svc.Run
+	isInteractive, err := svc.IsAnInteractiveSession()
+	if err != nil {
+		log.Error("Unable to ascertain if running as an Interactive Session: %v", err)
+		return
+	}
+	if isInteractive {
+		run = debug.Run
+	}
+	go run(WindowsServiceName, g)
+}
+
+// Execute makes gracefulManager implement svc.Handler
+func (g *gracefulManager) Execute(args []string, changes <-chan svc.ChangeRequest, status chan<- svc.Status) (svcSpecificEC bool, exitCode uint32) {
+	status <- svc.Status{State: svc.StartPending}
+
+	// Now need to wait for everything to start...
+	if !g.awaitServer(10 * time.Second) {
+		return false, 1
+	}
+
+	// We need to implement some way of svc.AcceptParamChange/svc.ParamChange
+	status <- svc.Status{
+		State:   svc.Running,
+		Accepts: svc.AcceptStop | svc.AcceptShutdown | acceptHammerCode,
+	}
+
+	waitTime := 30 * time.Second
+
+loop:
+	for change := range changes {
+		switch change.Cmd {
+		case svc.Interrogate:
+			status <- change.CurrentStatus
+		case svc.Stop, svc.Shutdown:
+			g.doShutdown()
+			waitTime += setting.GracefulHammerTime
+			break loop
+		case hammerCode:
+			g.doShutdown()
+			g.doHammerTime(0 *time.Second)
+			break loop
+		default:
+			log.Debug("Unexpected control request: %v", change.Cmd)
+		}
+	}
+
+	status <- svc.Status{
+		State: svc.StopPending,
+		WaitHint: uint32(waitTime/time.Millisecond),
+	}
+
+hammerLoop:
+	for {
+		select {
+		case change := <-changes:
+			switch change.Cmd {
+			case svc.Interrogate:
+				status <- change.CurrentStatus
+			case svc.Stop, svc.Shutdown, hammerCmd:
+				g.doHammerTime(0 * time.Second)
```

Comment:
I may be wrong about the usage of these timers, but it looks like 1 second is quite low. My PC sometimes freezes for 3~5 seconds while opening lightweight programs like notepad, and _it has an SSD_.

Reply:
So this is in the hammerLoop - it's only if you send a second svc.Stop/etc. to the server once you've already told it to shutdown. 

As far the windows documentation goes (https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-controlservice) you shouldn't send any more controls to a service you've already told to stop, but we're basically allowing this here to allow users to forcibly hammer if they want Gitea to just stop already.

Reply:
Mmm... Does Windows actually allow that? I mean, you may find out that Windows will ignore extra commands if the service is in the stopping state. I thought that hammer time was a goroutine that would automatically trigger after a timeout.

Note: you may also want to reconsider the 10 second delay in `manager_windows.go:(g *gracefulManager) Execute()`. Gitea startup may trigger the database startup and other services' so it may very well take longer than 10 seconds.

Also, you need to add a `// You can't touch this!` comment somewhere in the function, otherwise Hammer Time will not work correctly. 😁

Reply:
Oh HammerTime happens by default, this is just to forcibly hammer if you want to shutdown quicker. If windows doesn't hand these over then that's fine we can remove it - I have no idea if it would or not.

re: awaitServer - any ideas of a reasonable value? I guess we could just add another setting.

Reply:
> 
> 
> re: awaitServer - any ideas of a reasonable value? I guess we could just add another setting.

~~I'd go at least for one minute.~~ As I said, Gitea's start may trigger the database start, which involves several services in the case of MSSQL, which in turn might trigger a database recovery operation, etc etc. We're talking of system start, here. You may not remember, but old mechanical disk backed Windows can easily make you wait 5 minutes before you can log in.

EDIT (before posting 😁): it turns out that the default Windows timeout for a service start is 30 seconds, [but users can change that in the registry](https://serverfault.com/questions/622432/how-do-i-increase-windows-service-startup-timeout/622435). What happens after the 30 seconds? I don't know. So perhaps a better timeout for us would be 25 seconds?

Reply:
I had it at 30 seconds originally.

Reply:
> 
> 
> I had it at 30 seconds originally.

Uhmmm... 🤔 I don't remember, sorry.

Reply:
I'll up it back to 25